File: Tyche.Android/session/src/main/java/com/felipearpa/tyche/session/authentication/application/SignInWithEmailAndPasswordUseCase.kt
package com.felipearpa.tyche.session.authentication.application

import com.felipearpa.tyche.core.type.Email
import com.felipearpa.tyche.session.AccountBundle
import com.felipearpa.tyche.session.AccountStorage
import com.felipearpa.tyche.session.authentication.domain.AccountLink
import com.felipearpa.tyche.session.authentication.domain.AuthenticationRepository

class SignInWithEmailAndPasswordUseCase(
    private val authenticationRepository: AuthenticationRepository,
    private val accountStorage: AccountStorage
) {
    suspend fun execute(email: Email, password: String): Result<AccountBundle> {
        val externalAccountId =
            authenticationRepository.signInWithEmailAndPassword(
                email = email.value,
                password = password
            )
                .onFailure { exception -> return Result.failure(exception) }
                .getOrNull()!!

        val accountBundle = authenticationRepository.linkAccount(
            accountLink = AccountLink(
                email = email,
                externalAccountId = externalAccountId
            )
        )
            .onFailure { exception -> return Result.failure(exception) }
            .getOrNull()!!

        accountStorage.store(accountBundle = accountBundle)

        return Result.success(accountBundle)
    }
}

File: Tyche.iOS/Session/Sources/Session/Authentication/Application/SignInWithEmailAndPasswordUseCase.swift
import Core

public class SignInWithEmailAndPasswordUseCase {
    let authenticationRepository: AuthenticationRepository
    let accountStorage: AccountStorage
    
    init(authenticationRepository: AuthenticationRepository, accountStorage: AccountStorage) {
        self.authenticationRepository = authenticationRepository
        self.accountStorage = accountStorage
    }
    
    public func execute(email: Email, password: String) async -> Result<AccountBundle, Error> {
        let signInResult = await authenticationRepository.signInWithEmailAndPassword(email: email.value, password: password)
        guard case .success(let externalAccountId) = signInResult else {
            return Result.failure(signInResult.errorOrNil()!)
        }
        
        let linkAccountResult = await authenticationRepository.linkAccount(
            accountLink: AccountLink(email: email, externalAccountId: externalAccountId)
        )
        guard case .success(let accountBundle) = linkAccountResult else {
            return Result.failure(linkAccountResult.errorOrNil()!)
        }

        do {
            try await accountStorage.store(accountBundle: accountBundle)
        } catch {
            return Result.failure(error)
        }
        
        return Result.success(accountBundle)
    }
}
